crates/data-plane-controller/src/controller.rs

```diff
@@ -0,0 +1,875 @@
+                .arg("--stack")
+                .arg(&state.stack_name)
+                .arg("--json")
+                .arg("--non-interactive")
+                .arg("--show-secrets")
+                .arg("--cwd")
+                .arg(&checkout.path())
+                .env("PULUMI_BACKEND_URL", self.state_backend.as_str())
+                .env("VIRTUAL_ENV", checkout.path().join("venv")),
+        )
+        .await?;
+
+        if !output.status.success() {
+            anyhow::bail!(
+                "pulumi stack output failed: {}",
+                String::from_utf8_lossy(&output.stderr),
+            );
+        }
+
+        let stack::PulumiExports {
+            ansible,
+            mut control,
+        } = serde_json::from_slice(&output.stdout).context("failed to parse pulumi output")?;
+
+        // Install Ansible requirements.
+        () = run_cmd(
+            async_process::Command::new(checkout.path().join("venv/bin/ansible-galaxy"))
+                .arg("install")
+                .arg("--role-file")
+                .arg("requirements.yml")
+                .current_dir(checkout.path()),
+            "ansible-install",
+            &self.logs_tx,
+            state.logs_token,
+        )
+        .await?;
+
+        // Write out Ansible inventory.
+        std::fs::write(
+            checkout.path().join("ansible-inventory.json"),
+            serde_json::to_vec_pretty(&ansible).context("failed to serialize ansible inventory")?,
+        )
+        .context("failed to write ansible inventory")?;
+
+        // Write out Ansible SSH key and set it to 0600.
+        // Ansible is sensitive about their being a trailing newline.
+        let ssh_key_path = checkout.path().join("ansible-ssh.key");
+        let mut ssh_key = std::mem::take(&mut control.ssh_key);
+
+        if !ssh_key.ends_with("\n") {
+            ssh_key.push('\n');
+        }
+
+        std::fs::write(&ssh_key_path, ssh_key).context("failed to write ansible SSH key")?;
+        std::fs::set_permissions(
+            ssh_key_path,
+            std::os::unix::fs::PermissionsExt::from_mode(0o600),
+        )
+        .context("failed to set permissions of ansible SSH key")?;
+
+        // Run the Ansible playbook.
+        () = run_cmd(
+            async_process::Command::new(checkout.path().join("venv/bin/ansible-playbook"))
+                .arg("data-plane.ansible.yaml")
+                .current_dir(checkout.path())
+                .env("ANSIBLE_FORCE_COLOR", "1"),
+            "ansible-playbook",
+            &self.logs_tx,
+            state.logs_token,
+        )
+        .await?;
+
+        // Now that we've completed Ansible, all deployments are current and we
+        // can prune empty deployments.
+        state
+            .stack
+            .config
+            .model
+            .deployments
+            .retain_mut(|deployment| {
+                deployment.current = deployment.desired;
+                deployment.current != 0
```

Comment:
nit: is this two lines for better readability? i.e. to show that now we have attained `desired` number of replicas before the retain condition

Asking since this could be simplified to:
```suggestion
                deployment.desired != 0
```

Reply:
No, it's important / semantic. If desired is `3`, we need to update `current` from (for example) `1 => 3` in cases where the deployment is retained. We're doing an interior state mutation which is persisted back to the task row.

Reply:
Ah I see, the `Outcome` returned updates the data_planes row after each run of the controller, makes sense now 👍🏽